tests/Unit/Utilities/Test_Functional.cpp

```diff
@@ -3,16 +3,153 @@
+                  T1, std::remove_reference<decltype(                        \
+                          std::declval<T1>() TESTOP std::declval<T2>())>>> = \
+                  nullptr>                                                   \
+    void operator()(UniformGen gen, UniformCustomDistribution<T1> dist1,     \
+                    UniformCustomDistribution<T2> dist2) {                   \
+      T1 val1 = dist1(gen);                                                  \
+      T1 val1_copy = val1;                                                   \
+      T2 val2 = dist2(gen);                                                  \
+      auto result = val1_copy TESTOP val2;                                   \
+      STRUCTNAME<>{}(val1, val2); /*NOLINT*/                                 \
+      CHECK(val1 == result);                                                 \
+    }                                                                        \
+    template <typename T1, typename T2, typename UniformGen,                 \
+              Requires<not cpp17::is_same_v<                                 \
+                  T1, std::remove_reference<decltype(                        \
+                          std::declval<T1>() TESTOP std::declval<T2>())>>> = \
+                  nullptr>                                                   \
+    void operator()(UniformGen /*gen*/,                                      \
+                    UniformCustomDistribution<T1> /*dist1*/,                 \
+                    UniformCustomDistribution<T2> /*dist2*/) {}              \
+  }
+
+MAKE_UNARY_TEST(Abs, abs);
+MAKE_UNARY_TEST(Acos, acos);
+MAKE_UNARY_TEST(Acosh, acosh);
+MAKE_UNARY_TEST(Asin, asin);
+MAKE_UNARY_TEST(Asinh, asinh);
+MAKE_UNARY_TEST(Atan, atan);
+MAKE_UNARY_TEST(Atanh, atanh);
+MAKE_UNARY_TEST(Cbrt, cbrt);
+MAKE_UNARY_TEST(Cos, cos);
+MAKE_UNARY_TEST(Cosh, cosh);
+MAKE_UNARY_TEST(Erf, erf);
+MAKE_UNARY_TEST(Exp, exp);
+MAKE_UNARY_TEST(Exp10, exp10);
+MAKE_UNARY_TEST(Exp2, exp2);
+MAKE_UNARY_TEST(Fabs, fabs);
+MAKE_UNARY_TEST(Imag, imag);
+MAKE_UNARY_TEST(InvCbrt, invcbrt);
+MAKE_UNARY_TEST(InvSqrt, invsqrt);
+MAKE_UNARY_TEST(Log, log);
+MAKE_UNARY_TEST(Log10, log10);
+MAKE_UNARY_TEST(Log2, log2);
+MAKE_UNARY_TEST(Real, real);
+MAKE_UNARY_TEST(Sin, sin);
+MAKE_UNARY_TEST(Sinh, sinh);
+MAKE_UNARY_TEST(StepFunctionStruct, step_function);
+MAKE_UNARY_TEST(Sqrt, sqrt);
+MAKE_UNARY_TEST(Tan, tan);
+MAKE_UNARY_TEST(Tanh, tanh);
+
+MAKE_BINARY_TEST(Atan2, atan2);
+MAKE_BINARY_TEST(Hypot, hypot);
+MAKE_BINARY_TEST(Pow, pow);
+
+MAKE_BINARY_OP_TEST(Divides, /);
+MAKE_BINARY_OP_TEST(Minus, -);
+MAKE_BINARY_OP_TEST(Multiplies, *);
+MAKE_BINARY_OP_TEST(Plus, +);
+
+MAKE_BINARY_INPLACE_TEST(DivAssign, /=, /);
+MAKE_BINARY_INPLACE_TEST(MinusAssign, -=, -);
+MAKE_BINARY_INPLACE_TEST(MultAssign, *=, *);
+MAKE_BINARY_INPLACE_TEST(PlusAssign, +=, +);
+
+using RealTypeList = tmpl::list<float, double>;
```

Comment:
Would testing `int` be as easy as adding it to this list? If so I'd be in favor of testing int and long as well. Same with complex<double> and int and long

Reply:
yep! I can just add to the list. Interesting point that I discovered doing so is that many functions just fail on complex<integer types>. I think we should just not try to support that for now.

Reply:
Okay, let's avoid them :) I personally don't know when we would even want a complex integer :/